tooling/lsp/src/requests/workspace_symbol.rs

```diff
@@ -0,0 +1,300 @@
+use fuzzy_matcher::{FuzzyMatcher, skim::SkimMatcherV2};
+use lsp_types::{SymbolKind, Url, WorkspaceSymbol, WorkspaceSymbolParams, WorkspaceSymbolResponse};
+use nargo::{insert_all_files_under_path_into_file_manager, parse_all};
+use noirc_errors::{Location, Span};
+use noirc_frontend::{
+    ast::{
+        Ident, LetStatement, NoirEnumeration, NoirFunction, NoirStruct, NoirTrait, NoirTraitImpl,
+        NoirTypeAlias, Pattern, TraitImplItemKind, TraitItem, TypeImpl, Visitor,
+    },
+    parser::ParsedSubModule,
+};
+
+use crate::{LspState, source_code_overrides};
+
+use super::to_lsp_location;
+
+pub(crate) fn on_workspace_symbol_request(
+    state: &mut LspState,
+    params: WorkspaceSymbolParams,
+) -> impl Future<Output = Result<Option<WorkspaceSymbolResponse>, ResponseError>> + use<> {
+    let Some(root_path) = state.root_path.clone() else {
+        return future::ready(Ok(None));
+    };
+
+    let overrides = source_code_overrides(&state.input_files);
+
+    // Prepare a FileManager for files we'll parse, to extract symbols from
+    let mut file_manager = FileManager::new(root_path.as_path());
+
+    let cache = &mut state.workspace_symbol_cache;
+
+    // If the cache is not initialized yet, put all files in the workspace in the FileManager
+    if !cache.initialized {
+        insert_all_files_under_path_into_file_manager(&mut file_manager, &root_path, &overrides);
+        cache.initialized = true;
+    }
+
+    // Then add files that we need to re-process
+    for path in std::mem::take(&mut cache.paths_to_reprocess) {
+        if !path.exists() {
+            continue;
+        }
+
+        if let Some(source) = overrides.get(path.as_path()) {
+            file_manager.add_file_with_source(path.as_path(), source.to_string());
+        } else {
+            let source = std::fs::read_to_string(path.as_path())
+                .unwrap_or_else(|_| panic!("could not read file {:?} into string", path));
+            file_manager.add_file_with_source(path.as_path(), source);
+        }
+    }
+
+    // Note: what happens if a file is deleted? We don't get notifications when a file is deleted
+    // so we might return symbols for a file that doesn't exist. However, VSCode seems to notice
+    // the file doesn't exist and simply doesn't show the symbol. What if the file is re-created?
+    // In that case we do get a notification so we'll reprocess that file.
+
+    // Parse all files for which we don't know their symbols yet,
+    // figure out the symbols and store them in the cache.
+    let parsed_files = parse_all(&file_manager);
+    for (file_id, (parsed_module, _)) in parsed_files {
+        let path = file_manager.path(file_id).unwrap().to_path_buf();
+        let mut symbols = Vec::new();
+        let mut gatherer = WorkspaceSymboGatherer::new(&mut symbols, file_manager.as_file_map());
+        parsed_module.accept(&mut gatherer);
+        cache.symbols_per_path.insert(path, gatherer.symbols.clone());
```

Comment:
```suggestion
        cache.symbols_per_path.insert(path, symbols);
```
It looks like the gatherer doesn't own it, and won't need it any more either.

Reply:
Thanks! I also moved `symbols` to be owned by it.